<div align="center">
<h1>Jurnal Praktikum Dasar Kecerdasan Artifisial</h1>
<h3>Modul 2: Pengenalan Library NumPy</h3>
</div>

### Tujuan Praktikum
- Memahami konsep dasar array multidimensi dan bagaimana NumPy dapat digunakan untuk memanipulasi data secara efisien.
- Menguasai penggunaan fungsi dan operasi dasar NumPy, seperti pembuatan array, reshaping, indexing, serta pengolahan statistik.
- Mengaplikasikan NumPy dalam pemecahan masalah numerik dan pemodelan data untuk keperluan kecerdasan buatan dan analisis data.

**Jangan lupa untuk selalu import Library NumPy memulai sesi baru (run time) atau script Python yang memerlukan penggunaan library NumPy.**

In [1]:
import numpy as np

### 1. Definisi Array dan Array NumPy (Bobot Nilai: `5`)

**a.** Lengkapi potongan kode di bawah ini untuk membuat array Python biasa dan array NumPy, kemudian bandingkan perbedaan bentuknya.

In [ ]:
# Inisialisasi array Python biasa
array_python = [] # Ganti nilai None dengan inisialisai array Python biasa

# Inisialisasi array NumPy
array_numpy = np.array([]) # Ganti nilai None dengan inisialisasi array NumPy

# Cek tipe variabel
print("Tipe array_python:", type(array_python))
print("Tipe array_numpy:", type(array_numpy))

Tipe array_python: <class 'list'>
Tipe array_numpy: <class 'numpy.ndarray'>


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |'Tipe array_python: <class 'list'>'<br>'Tipe array_numpy: <class 'numpy.ndarray'>'|

**b.** Buat program yang menerima input dari pengguna berupa daftar bilangan, kemudian bentuk array Python biasa dan array NumPy dari daftar tersebut.

In [4]:
# Input dari user
data = input("Masukkan daftar bilangan, pisahkan dengan koma: ").split(",")

# Masukkan data ke array Python biasa
array_python = data

# Masukkan data ke array NumPy
array_numpy = np.array(data) 

# Tampilkan hasil
print("Array Python:", array_python)
print("Array NumPy:", array_numpy)

Array Python: ['1', '2', '3', '4']
Array NumPy: ['1' '2' '3' '4']


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|1,2,3,4|Array Python: [1, 2, 3, 4]<br>Array NumPy: [1 2 3 4]|
|10,20,30|Array Python: [10, 20, 30]<br>Array NumPy: [10 20 30]|

### 2. Atribut Array NumPy (Bobot Nilai: `10`)

Lengkapilah kode berikut untuk memeriksa atribut-atribut dimensi, bentuk, ukuran, dan tipe data dari array NumPy yang telah dibuat. Tujuan dari tugas ini adalah memahami bagaimana mendapatkan dimensi, bentuk, ukuran, dan tipe data dari array NumPy.

In [9]:
# Inisialisasi array NumPy
array_numpy = np.array([[1, 2, 3], [4, 5, 6]])

# Periksa jumlah dimensi 
jumlah_dimensi = array_numpy.ndim

# Periksa bentuk array (shape)
bentuk_array = array_numpy.shape 

# Periksa ukuran array (size)
ukuran_array = array_numpy.size 

# Periksa tipe data array (dtype)
tipe_data = array_numpy.dtype  

print("Jumlah dimensi:", jumlah_dimensi)
print("Bentuk array:", bentuk_array)
print("Ukuran array:", ukuran_array)
print("Tipe data array:", tipe_data)

Jumlah dimensi: 2
Bentuk array: (2, 3)
Ukuran array: 6
Tipe data array: int64


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |'Jumlah dimensi: 2'<br>'Bentuk array: (2, 3)'<br>'Ukuran array: 6'<br>'Tipe data array: int64'|

### 3. Deklarasi Array dengan NumPy (Bobot Nilai: `10`)

**a.** Inisialisasikan sebuah array 2D 2x3 yang berisi nilai nol.

In [32]:
# Inisialisasi array dengan nilai nol
array = np.array([[0,0,0],[0,0,0]])  

print(array)

[[0 0 0]
 [0 0 0]]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[[0. 0. 0.]<br>[0. 0. 0.]]|

**b.** Inisialisasikan sebuah array 3D 3x2x4 yang berisi nilai satu.

In [38]:
# Inisialisasi array dengan nilai satu
array = np.array([[[1,1,1,1],[1,1,1,1]],[[1,1,1,1],[1,1,1,1]],[[1,1,1,1],[1,1,1,1]]])  

print(array)

[[[1 1 1 1]
  [1 1 1 1]]

 [[1 1 1 1]
  [1 1 1 1]]

 [[1 1 1 1]
  [1 1 1 1]]]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[[[1. 1. 1. 1.]<br>  [1. 1. 1. 1.]]<br><br> [[1. 1. 1. 1.]<br>  [1. 1. 1. 1.]]<br><br> [[1. 1. 1. 1.]<br>  [1. 1. 1. 1.]]]|

**c.** Inisialisasikan sebuah array 1D berukuran 5 dengan nilai acak.

In [28]:
# Inisialisasi array dengan nilai acak
array = np.random.random(size=5)

print(array)

[0.26078017 0.92589687 0.46822825 0.25707442 0.14865751]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[6.23042070e-307   4.67296746e-307   1.69121096e-306   6.11557788e-312   3.33510022e-317]<br>*Catatan: Angka akan tergenerate secara random*|

**d.** Inisialisasikan sebuah array 1D dengan range nilai dari 0 hingga 10.

In [39]:
# Inisialisasi array dengan range nilai
array = np.arange(0,10)

print(array)

[0 1 2 3 4 5 6 7 8 9]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[0 1 2 3 4 5 6 7 8 9]|

**e.** Inisialisasikan sebuah array 1D yang memiliki nilai yang terdistribusi merata antara 0 hingga 1 dengan 5 elemen

In [40]:
# Inisialisasi array dengan linspace
array = np.linspace(0,1,5) 

print(array)

[0.   0.25 0.5  0.75 1.  ]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[0. 0.25 0.5 0.75 1. ]|

**f.** Buatlah array 2D berukuran 2x3 dengan semua elemen diisi dengan angka 7

In [41]:
# Inisialisasi array dengan nilai tetap
array = np.full((2,3),7) 

print(array)

[[7 7 7]
 [7 7 7]]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[[7 7 7]<br> [7 7 7]]|

**g.** Buatlah sebuah array 1D dengan 4 nilai acak antara 0 dan 1

In [42]:
# Inisialisasi array dengan nilai acak
array = np.random.rand(4)

print(array)

[0.24533794 0.09124883 0.96775194 0.35430192]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[0.67039964 0.38240816 0.9551951  0.07901245]<br>*Catatan: Angka akan tergenerate secara random*|

**h.** Buatlah array 2D berukuran 3x3 yang berisi bilangan acak bulat antara 1 dan 10

In [50]:
# Inisialisasi array dengan bilangan bulat acak
array = np.random.randint(1,11,(3,3))

print(array)

[[4 3 7]
 [7 9 3]
 [5 8 8]]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[[5 2 9] <br>[1 8 3] <br>[7 6 4]]<br>*Catatan: Angka akan tergenerate secara random*|

**i.** Buatlah array 2D berukuran 3x3 yang isinya merupakan menghasilkan nilai dari perkalian baris dan kolomnya

In [51]:
# Inisialisasi array menggunakan fungsi


# Buat function perkalian dari dua bilangan
def perkalian(x,y):
    return x * y

array = np.fromfunction(perkalian,(3,3),dtype=int) # Ganti None dengan sintaks deklarasi array yang sesuai

print(array)

[[0 0 0]
 [0 1 2]
 [0 2 4]]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[[0. 0. 0.] <br>[0. 1. 2.] <br>[0. 2. 4.]]|

### 4. Penambahan Elemen Array (Bobot Nilai:`5`)

**a.** Anda bekerja dengan data mahasiswa yang disimpan dalam array NumPy. Data ini terdiri dari nama-nama mahasiswa dari dua kelas berbeda. Tugas Anda adalah menggabungkan data dari kedua kelas tersebut ke dalam satu array.

In [52]:
# Data dari kelas A dan kelas B
kelas_A = np.array(['Afeef', 'Rafly', 'Arya'])
kelas_B = np.array(['Reza', 'Valen', 'Iksan'])

# Gabungkan kedua array 
data_mahasiswa = np.concatenate((kelas_A,kelas_B)) 
print(data_mahasiswa)

['Afeef' 'Rafly' 'Arya' 'Reza' 'Valen' 'Iksan']


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |['Afeef' 'Rafly' 'Arya' 'Reza' 'Valen' 'Iksan']|

**b.** Dua matakuliah memerlukan penggabungan data nilai untuk analisis lebih lanjut. Nilai kedua matkul tersebut disimpan dalam array terpisah. Gabungkan kedua array ini menjadi satu. Usahakan gunakan sintaks penggabungan array yang berbeda dengan soal sebelumnya. 

In [54]:
# Nilai matkul A dan B
nilai_A = np.array([75, 80, 85])
nilai_B = np.array([70, 90, 95])

# Gabungkan kedua array
nilai_total = np.append(nilai_A,nilai_B)

print(nilai_total)

[75 80 85 70 90 95]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[75 80 85 70 90 95]|

**c.** Anda diminta untuk menambahkan nilai ujian tengah semester (UTS) pada posisi tertentu dalam array nilai mahasiswa.

In [55]:
# Nilai akhir mahasiswa
nilai = np.array([85, 90, 95])

# Sisipkan nilai UTS (88) di posisi ke-2
nilai_dengan_uts = np.insert(nilai,1,88) 
print(nilai_dengan_uts)

[85 88 90 95]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[85 88 90 95]|

**d.** Anda diminta untuk menggabungkan dua array nilai ke dalam sebuah array 2D untuk mempermudah analisis lebih lanjut. Usahakan gunakan sintaks penggabungan array yang berbeda dengan soal sebelumnya.

In [64]:
# Nilai dari dua kelas
nilai_kelas_A = np.array([80, 85, 90])
nilai_kelas_B = np.array([75, 88, 92])

# Gabungkan menjadi array 2D 
nilai_gabungan = np.stack((nilai_kelas_A,nilai_kelas_B)) 
print(nilai_gabungan)

[[80 85 90]
 [75 88 92]]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[[80 85 90]<br> [75 88 92]]|

### 5. Penghapusan Elemen Array (Bobot Nilai: `10`)

**a.** Seorang peneliti ingin membersihkan data yang mengandung elemen salah dalam sebuah array 1D yang berisi nilai eksperimen. Hapus elemen sesuai dengan index input si peneliti.

In [66]:
# Array berisi nilai eksperimen
nilai_eksperimen = np.array([10, 15, 18, 22])

# Hapus elemen di indeks yang sesuai
nilai_bersih = np.delete(nilai_eksperimen,1)
print(nilai_bersih)

[10 18 22]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|1|[10 18 22]|
|3|[10 15 18]|

**b.** Anda bekerja dengan data hasil tes dari beberapa kelompok siswa yang disimpan dalam array 2D. Namun, data dari kelompok ke-3 dan semua siswa pertama harus dihapus karena salah pencatatan. Hapus baris ketiga dan kolom pertama dari array tersebut.

In [126]:
# Data hasil tes (dalam 2D array)
hasil_tes = np.array([[85, 90, 88],
                      [78, 82, 84],
                      [93, 89, 92],
                      [75, 80, 85]])

print("Masukkan nomor kelompok dan urutan siswa, dengan dipisahkan oleh koma (,)")

# Simpan nomor kelompok (baris) dan urutan siswa (kolom) dalam list
target_hapus = input()
baris, kolom = map(int, target_hapus.split(','))

# Hapus data dari sesuai dengan target hapus
hasil_tes_baru = np.delete(hasil_tes, baris-1, axis=0)
hasil_tes_baru = np.delete(hasil_tes_baru, kolom-1, axis=1)

print(hasil_tes_baru)

Masukkan nomor kelompok dan urutan siswa, dengan dipisahkan oleh koma (,)
[[90 88]
 [82 84]
 [80 85]]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|3,1|[[90 88]<br>[82 84]<br>[80 85]]|
|4,2|[[85 88]<br>[78 84]<br>[93 92]]|

**c.** Seorang ilmuwan ingin menghapus semua elemen yang berada di bawah batas nilai dalam array yang berisi data eksperimen.

In [75]:
# Data eksperimen
data = np.array([10, 3, 15, 7, 22, 17])

# Menerima input batas nilai
batas_nilai = int(input())

# Menghapus elemen dibawah batas nilai dari array
data_bersih = np.delete(data,data<batas_nilai)

print(data_bersih)

[15 22 17]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|15|[15. 22. 17.]|
|8|[10. 15. 22. 17.]|

### 6. Pengurutan Elemen Array (Bobot Nilai: `5`)

**a.** Seorang petugas inventaris ingin menyusun harga barang-barang yang sudah diinput ke dalam sistem. 

In [73]:
# Data harga barang
harga_barang = np.array([35000, 25000, 50000, 45000, 30000])

# Urutkan harga barang
harga_urut = np.sort(harga_barang)

print(harga_urut)

[25000 30000 35000 45000 50000]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[25000 30000 35000 45000 50000]|

**b.** Anda memiliki data hasil tes siswa dalam array 2D, di mana setiap baris adalah skor dari satu siswa.

In [74]:
# Data hasil tes siswa
skor_siswa = np.array([[85, 78, 90], [92, 89, 85], [76, 84, 80]])

# Urutkan skor setiap siswa
skor_urut = np.sort(skor_siswa)

print(skor_urut)

[[78 85 90]
 [85 89 92]
 [76 80 84]]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[[78 85 90]<br> [85 89 92]<br> [76 80 84]]|

**c.** Tim peneliti ingin mengurutkan semua data dalam array multidimensi berdasarkan nilainya (secara keseluruhan, bukan hanya per baris atau kolom). 

In [93]:
# Data eksperimen
data = np.array([[9, 4, 6], [3, 7, 1]])

# Urutkan semua elemen menjadi 1 array
data_urut = np.sort(data.flatten())
# data_urut = np.sort(data.ravel()) #hasilnya sama aja

print(data_urut)

[1 3 4 6 7 9]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |[1 3 4 6 7 9]|

### 7. Indexing dan Slicing Array (Bobot Nilai: `10`)

**a.** Seorang peneliti memiliki data suhu harian selama satu bulan dalam bentuk array (30 hari). Peneliti ingin mengetahui suhu pada hari ke-5, hari ke-10, hari ke-25, dan hari ke-30. Setelah mendapatkan suhu-suhu tersebut, lakukan perhitungan rata-rata.

In [107]:
# Data suhu selama 30 hari dengan range suhu 20 sampai 40 (buat menggunakan inisialisasi random integer)
suhu = np.random.randint(20,41,size=30)

# Ambil suhu pada hari ke-5, 10, 25, dan 30
suhu_tertentu = suhu[[4,9,24,29]]

# Hitung rata-rata suhu
rata_rata = np.mean(suhu)


print(f"Suhu hari tertentu: {suhu_tertentu}")
print(f"Rata-rata: {rata_rata}")

Suhu hari tertentu: [35 29 40 21]
Rata-rata: 29.433333333333334


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Suhu hari tertentu: [31 31 25 25]<br>Rata-rata: 28.0<br>*Catatan: Angka akan tergenerate secara random*|

**b.** Seorang ilmuwan mencatat jumlah bakteri selama 10 hari dalam sebuah array. Namun, ia hanya ingin mengambil data pertumbuhan bakteri setiap 2 hari (hari ke-1, 3, 5, dan seterusnya) dan menghitung total pertumbuhannya.

In [112]:
# Data pertumbuhan bakteri selama 10 hari
pertumbuhan_bakteri = np.array([100, 120, 240, 130, 150, 110, 220, 175, 160, 200])

# Ambil data setiap 2 hari
data_tertentu = pertumbuhan_bakteri[::2]

# Hitung total pertumbuhan
total_pertumbuhan = np.sum(data_tertentu)

print(f"Data bakteri: {data_tertentu}")
print(f"Total pertumbuhan: {total_pertumbuhan}")

Data bakteri: [100 240 150 220 160]
Total pertumbuhan: 870


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Data bakteri: [100 240 150 220 160]<br>Total pertumbuhan: 870|

**c.** Anda memiliki array berisi data stok barang di gudang. Anda ingin menghapus semua barang yang stoknya kurang dari 50 unit, kemudian hitung total stok barang yang tersisa. Gunakan boolean indexing untuk menyaring data dan hitung totalnya.

In [113]:
# Data stok barang
stok_barang = np.array([41, 51, 75, 57, 92, 13, 35, 64, 24, 66])

# Filter barang dengan stok >= 50
stok_tersisa = stok_barang[stok_barang>=50]

# Hitung total stok tersisa
total_stok = np.sum(stok_tersisa)

print(f"Stok barang tersisa: {stok_tersisa}")
print(f"Total stok tersisa: {total_stok}")

Stok barang tersisa: [51 75 57 92 64 66]
Total stok tersisa: 405


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Stok barang tersisa: [51 75 57 92 64 66]<br>Total stok tersisa: 405|

**d.** Anda memiliki data nilai siswa untuk 5 mata pelajaran yang disimpan dalam array. Guru ingin melihat nilai siswa ke-3 untuk mata pelajaran ke-2, ke-4, dan ke-5. Buatlah kode yang dapat menampilkan data tersebut.

In [114]:
# Data nilai siswa untuk 5 mata pelajaran
nilai = np.array([[80, 75, 85, 90, 95],
                  [70, 85, 80, 75, 90],
                  [88, 92, 77, 80, 85],
                  [65, 70, 85, 95, 100]])

# Ambil nilai siswa ke-3 untuk pelajaran ke-2, ke-4, dan ke-5
nilai_tertentu = nilai[2,[1,3,4]]

print(f"Nilai siswa: {nilai_tertentu}")

Nilai siswa: [92 80 85]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Nilai siswa: [92 80 85]|

### 8. Penyalinan (Copy) Array (Bobot Nilai: `5`)

**a.** Seorang ilmuwan data ingin melakukan analisis terhadap data eksperimen dengan mengubah data aslinya. Buat salinan "shallow copy" dari array data asli, lalu ubah elemen tertentu pada indeks tertentu dari data yang telah disalin.

In [5]:
# Data eksperimen asli
data_asli = np.array([10, 15, 20, 25, 30])

# Buat shallow copy
data_copy = data_asli.copy()

# Masukkan indeks elemen yang ingin diubah dan data baru dengan dipisahkan dengan koma
indeks_perubahan = 1
data_baru = 100

# Ubah data_copy pada indeks yang sesuai dengan data_baru
data_copy[indeks_perubahan] = data_baru

# Tampilkan data_copy dan data_asli
print("Data Copy:", data_copy)
print("Data Asli:", data_asli)

Data Copy: [ 10 100  20  25  30]
Data Asli: [10 15 20 25 30]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|1,100|Data Copy: [ 10 100  20  25  30]<br>Data Asli: [ 10 100  20  25  30]|

**b.** Seorang peneliti ingin melakukan eksperimen dengan data cuaca, dan memastikan bahwa perubahan data baru tidak mempengaruhi data aslinya. Buat "deep copy" dari array cuaca dan ubah data salinan tersebut.

In [6]:
# Data cuaca asli
data_cuaca = np.array([[30, 32, 35], [28, 27, 26], [33, 31, 29]])

# Buat deep copy dari data_cuaca
data_cuaca_copy = data_cuaca.copy()

# Ubah elemen di baris ke-2, kolom ke-3 data_cuaca_copy menjadi 40
data_cuaca_copy[1, 2] = 40

# Tampilkan data_cuaca_copy dan data_cuaca
print("Data Cuaca Copy:\n", data_cuaca_copy)
print("Data Cuaca Asli:\n", data_cuaca)

Data Cuaca Copy:
 [[30 32 35]
 [28 27 40]
 [33 31 29]]
Data Cuaca Asli:
 [[30 32 35]
 [28 27 26]
 [33 31 29]]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Data Cuaca Copy:<br> [[30 32 35]<br> [28 27 40]<br>[33 31 29]]<br>Data Cuaca Asli:<br> [[30 32 35]<br> [28 27 26]<br> [33 31 29]]|

### 9. Operasi Dasar Array (Bobot Nilai: `5`)

**a.** Seorang guru ingin mengetahui nilai tertinggi yang diperoleh siswa dari hasil ujian matematika. Tambahkan juga kode untuk menghitung selisih antara nilai maksimum dan nilai rata-rata ujian.

In [7]:
# Data nilai ujian siswa dalam array
nilai_ujian = np.array([75, 80, 90, 85, 60])

# Temukan nilai maksimum
nilai_maksimum = np.max(nilai_ujian)  

# Hitung nilai rata-rata
nilai_rata_rata = np.mean(nilai_ujian)

# Hitung selisih antara nilai maksimum dan rata-rata
selisih = nilai_maksimum - nilai_rata_rata  

# Tampilkan nilai maksimum dan selisihnya
print("Nilai Maksimum:", nilai_maksimum)
print("Selisih Maksimum dan Rata-rata:", selisih)

Nilai Maksimum: 90
Selisih Maksimum dan Rata-rata: 12.0


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Nilai Maksimum: 90<br> Selisih Maksimum dan Rata-rata: 12.0|

**b.** Seorang pelatih renang ingin mengetahui waktu renang tercepat dari setiap atlet. Hitung juga perbedaan waktu antara waktu tercepat dan waktu rata-rata.

In [17]:
# Data waktu renang atlet (dalam detik)
waktu_renang = np.array([50.5, 52.0, 48.7, 49.8, 51.2])

# Temukan waktu tercepat
waktu_tercepat = np.min(waktu_renang)

# Hitung waktu rata-rata
waktu_rata_rata = np.mean(waktu_renang) 

# Hitung perbedaan waktu tercepat dengan rata-rata
selisih_waktu = waktu_tercepat - waktu_rata_rata 

# Tampilkan waktu tercepat dan selisihnya
print("Waktu Tercepat:", waktu_tercepat)
print("Selisih Waktu Tercepat dan Rata-rata:", round(selisih_waktu,0))

Waktu Tercepat: 48.7
Selisih Waktu Tercepat dan Rata-rata: -2.0


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Waktu Tercepat: 48.7<br> Selisih Waktu Tercepat dan Rata-rata: -2.0|

**c.** Dalam sebuah kompetisi, seorang juri ingin mengetahui total skor yang diperoleh tim A selama lima putaran pertandingan. Tampilkan rata-rata skor per putaran.

In [18]:
# Data skor tim A dalam 5 putaran
skor_tim_A = np.array([85, 78, 90, 88, 76])

# Hitung total skor
total_skor = np.sum(skor_tim_A)  

# Hitung rata-rata skor per putaran
rata_rata_skor = np.mean(skor_tim_A)  

# Tampilkan total skor dan rata-ratanya
print("Total Skor:", total_skor)
print("Rata-rata Skor per Putaran:", rata_rata_skor)

Total Skor: 417
Rata-rata Skor per Putaran: 83.4


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Total Skor: 417<br> Rata-rata Skor per Putaran: 83.4|

**d.** Seorang analis pasar saham ingin mengetahui rata-rata harga penutupan saham dari lima hari terakhir. Hitung rata-rata harga penutupan, dan tambahkan kode untuk membandingkan harga penutupan terakhir dengan rata-rata.

In [19]:
# Data harga penutupan saham selama 5 hari 
harga_penutupan = np.array([150.5, 152.0, 148.7, 149.8, 151.2])

# Hitung rata-rata harga penutupan
rata_rata_harga = np.mean(harga_penutupan)  

# Bandingkan harga penutupan terakhir dengan rata-rata
harga_terakhir = harga_penutupan[-1] 
perbandingan = harga_terakhir - rata_rata_harga

# Tampilkan rata-rata harga dan hasil perbandingannya
print("Rata-rata Harga Penutupan:", rata_rata_harga)
print("Harga Penutupan Terakhir dibanding Rata-rata:", perbandingan)

Rata-rata Harga Penutupan: 150.44
Harga Penutupan Terakhir dibanding Rata-rata: 0.7599999999999909


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Rata-rata Harga Penutupan: 150.44<br> Harga Penutupan Terakhir dibanding Rata-rata: 0.7599999999999909|

### 10. Transpose dan Reshape Array (Bobot Nilai: `10`)

**a.** Seorang dosen ingin menampilkan data mahasiswa dalam format yang berbeda. Data saat ini disusun dalam bentuk baris, di mana setiap baris merepresentasikan mahasiswa. Anda diminta untuk membuat array tersebut sehingga data mahasiswa ditampilkan per kolom.

In [20]:
# Data mahasiswa (Nama, NIM, Jurusan)
data_mahasiswa = np.array([['Farah', 'Tiara', 'Imelda'], ['123', '456', '789'], ['IF', 'DS', 'RPL']])

# Ubah data_mahasiswa
data_1 = data_mahasiswa.T

print("Data mahasiswa:")
print(data_1)

Data mahasiswa:
[['Farah' '123' 'IF']
 ['Tiara' '456' 'DS']
 ['Imelda' '789' 'RPL']]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Data mahasiswa:<br> [['Farah' '123' 'IF']<br> ['Tiara' '456' 'DS']<br> ['Imelda' '789' 'RPL']]|

In [21]:
# Data nilai mahasiswa
nilai = np.array([80, 90, 85, 78, 92, 88])

# Bentuk ulang array nilai menjadi 2D dengan 3 kolom
nilai_1 = nilai.reshape(-1, 3)

print("Array nilai:")
print(nilai_1)

Array nilai:
[[80 90 85]
 [78 92 88]]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Array nilai:<br> [[80 90 85]<br> [78 92 88]]|

**c.** Sebuah perusahaan ingin menganalisis data penjualan dari berbagai cabang. Data saat ini ada dalam bentuk array 1D. Ubah array tersebut menjadi bentuk 2D dan transposisikan hasilnya agar data cabang dan produk lebih mudah dibaca.

In [22]:
# Data penjualan: [Cabang A, Cabang B, Cabang A, Cabang B, Cabang A, Cabang B]
penjualan = np.array([120, 150, 180, 100, 130, 170])

# Ubah array 
penjualan_2D = penjualan.reshape(-1, 2)
penjualan_1 = penjualan_2D.T

print("Data penjualan:")
print(penjualan_1)

Data penjualan:
[[120 180 130]
 [150 100 170]]


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Data penjualan:<br> [[120 180 130]<br>  [150 100 170]]|

### 11. Soal Cerita (Bobot Nilai: `25`)

Sebuah tim pelatih sedang mengevaluasi performa (indeks performa 6 - 10) tiga tim sepak bola selama 10 pertandingan. Statistik pertandingan disimpan dalam sebuah array yang merepresentasikan skor dari tiga tim (kolom) di setiap pertandingan (baris). Pelatih hanya tertarik pada pertandingan di mana rata-rata performa ketiga tim lebih dari atau sama dengan 7. Mereka ingin menganalisis lebih lanjut dengan menyalin data pertandingan tersebut, melakukan transpose pada data untuk melihat statistik dari perspektif tim, dan menghitung rata-rata performa tiap tim. Berikut adalah langkah-langkah yang perlu dilakukan:

- Ambil subset data dari pertandingan, yang dimana rata-rata performa 3 tim pada hari yang sama memiliki nilai lebih dari atau sama denga 7
- Buat salinan data untuk dianalisis lebih lanjut
- *Transpose array* hasil salinan statistik
- Ambil rata-rata performa tiap tim pada hari (baris) yang terdapat pada subset dari data pertandingan

In [25]:
# Buat array yang merepresentasikan statistik 10 pertandingan (10x3)
data_statistik = np.array([
    [8, 8, 7], 
    [10, 7, 9], 
    [8, 8, 6], 
    [6, 7, 6], 
    [6, 6, 9], 
    [9, 6, 6], 
    [6, 6, 7], 
    [6, 7, 6], 
    [8, 9, 10], 
    [6, 7, 6]
])

# You’re on Your Own, Kid. Good Luck :)
rata_rata_performa = np.mean(data_statistik, axis=1)
subset_data = data_statistik[rata_rata_performa >= 7] 

salinan_data = subset_data.copy()

transposed_data = salinan_data.T

rata_rata_tim_A = np.mean(transposed_data[0])
rata_rata_tim_B = np.mean(transposed_data[1])
rata_rata_tim_C = np.mean(transposed_data[2])

print("Data Statistik Asli:")
print(subset_data)
print("\nSubset Data Statistik:")
print(salinan_data)
print("\nData Setelah Di-Transpose:")
print(transposed_data)
print("\nRata - Rata Performa Team A:", rata_rata_tim_A)
print("Rata - Rata Performa Team B:", rata_rata_tim_B)
print("Rata - Rata Performa Team C:", rata_rata_tim_C)

Data Statistik Asli:
[[ 8  8  7]
 [10  7  9]
 [ 8  8  6]
 [ 6  6  9]
 [ 9  6  6]
 [ 8  9 10]]

Subset Data Statistik:
[[ 8  8  7]
 [10  7  9]
 [ 8  8  6]
 [ 6  6  9]
 [ 9  6  6]
 [ 8  9 10]]

Data Setelah Di-Transpose:
[[ 8 10  8  6  9  8]
 [ 8  7  8  6  6  9]
 [ 7  9  6  9  6 10]]

Rata - Rata Performa Team A: 8.166666666666666
Rata - Rata Performa Team B: 7.333333333333333
Rata - Rata Performa Team C: 7.833333333333333


*Test case* program

|**Input**|**Output**    |
|---------|--------------|
|         |Data Statistik Asli:<br>[[ 8  8  7]<br>[10  7  9]<br>[ 8  8  6]<br>[ 6  7  6]<br>[ 6  6  9]<br>[ 9  6  6]<br>[ 6  6  7]<br>[ 6  7  6]<br>[ 8  9 10]<br>[ 6  7  6]]<br>Subset Data Statistik:<br>[[ 8  8  7]<br>[10  7  9]<br>[ 8  8  6]<br>[ 6  6  9]<br>[ 9  6  6]<br>[ 8  9 10]]<br>Data Setelah Di-transpose:<br>[[ 8 10  8  6  9  8]<br>[ 8  7  8  6  6  9]<br>[ 7  9  6  9  6 10]]<br>Rata - Rata Performa Team A:  8.166666666666666<br>Rata - Rata Performa Team B:  7.333333333333333<br>Rata - Rata Performa Team C:  7.833333333333333|